segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -346,10 +352,9 @@ private void handleException(UUID writerId, long requestId, String segment, long
             log.warn("Bad attribute update by {} on segment {}.", writerId, segment, u);
             connection.send(new InvalidEventNumber(writerId, requestId, clientReplyStackTrace));
             connection.close();
-        } else if (u instanceof AuthenticationException) {
+        } else if (u instanceof TokenException) {
```

Comment:
Here we are replacing `AuthenticationException` with `TokenException`. `TokenException` extends `AuthException`, but I don't the relationship with `AuthenticationException`, could you clarify, please?

Reply:
The newly introduced TokenException can occur either while generating the token or when verifying it. In the case of Segment Store where this code lies, this exception can occur under conditions such as: 

* The token has expired (represented by `TokenExpiredException`, a subclass of `TokenException`)
* The token is invalid (represented by `InvalidTokenException`, a subclass of `TokenException`)
* The claim in the token doesn't match the authorization claim (represented by `InvalidClaimException`, a subclass of TokenException). 

The conditions that may lead to a `TokenException` include both authentication and authorization failures (apart from failure in token validation). `AuthenticationException` did not represent all those conditions, so the new `TokenException` and its subclasses were introduced rooted in `AuthException`. Technically, I could have used `AuthException` as well, but a `TokenException` is more representative of token-related failures. 

The segment store deals with delegation token verification, and therefore its `AppendProcessor` and `PravegaResourceProcessor` now throws TokenException back to the clients. 

Hope that helps.